### PR TITLE
Kernel memdomain fixes

### PR DESCRIPTION
**Pull request by ioannisg**
First commit is a bug fix, IMHO; it is about the end-boundaries of the partitions

The second commit is an acceleration of the sanity check for the added partitions, so the one-to-one partition checking is not performed twice, and added partitions are not sanity-checked against themselves

### CODE REVIEW

File: kernel/mem_domain.c [hunk 1]
```diff
@@ -22,7 +22,7 @@ static bool sane_partition(const struct k_mem_partition *part,
 	u32_t end;
 	u32_t i;
 
-	end = part->start + part->size;
+	end = part->start + part->size - 1;
```
Comment (andyross) [inline kernel/mem_domain.c]: Drive-by naming gripe: this probably should have been called "last", not "end".  It's the address of the last byte in the region.  Usually (but not always) "end" refers to the boundary, which is going to be the address of the byte after the region (i.e. so end = start + size, which is the intuitive "fact" that got mixed up here).

File: kernel/mem_domain.c [hunk 2]
```diff
@@ -81,7 +81,7 @@ void k_mem_domain_init(struct k_mem_domain *domain, u8_t num_parts,
 
 	key = irq_lock();
 
-	domain->num_partitions = num_parts;
+	domain->num_partitions = 0;
```
Comment (andyross) [inline kernel/mem_domain.c]: The cleanup and loop unification looks sane.    But.. I. don't see why this is going to optimize anything.  Nothing here is moving anything the compiler can't already see.  Is the generated code even any different?

**Review by andyross — APPROVED** [review]

File: kernel/mem_domain.c (hunk 1 above)
Comment (ioannisg) [reply kernel/mem_domain.c]: Right, so i can either
- replace "end" with "last" or "limit"
- or revert this change and modify line 42, so <, > become <=, >=
Believing that these are equivalent options, I am asking @andyross , what you would prefer :)

**Review by tyrantlizard77 — APPROVED** [review]
LGTM

File: kernel/mem_domain.c (hunk 2 above)
Comment (ioannisg) [reply kernel/mem_domain.c]: It is optimization, because:
Before:
Adding partitions is (i=0...N-1)
Sanity for each partition is (j=0...N-1) 

After:
Adding partitions is still (i=0...N-1)
Sanity check for each partition is j=0... i-1

So we end up with half of the loop iterations

File: kernel/mem_domain.c (hunk 1 above)
Comment (ioannisg) [reply kernel/mem_domain.c]: It's a valid point actually, not just a style thing. Fixed

**Comment by ioannisg** [comment]
@nashif I suppose this one can be merged, now.

**Review by AdithyaBaglody — APPROVED** [review]
